### DISCUSSION TITLE
User Stories ( Meetup / Login life cycle )

### DISCUSSION THREAD

**Discussion by eidolonFIRE** [eidolonFIRE/xcNav_web-DEPRECATED-#21] [Ideas]
## E) Spectating a flight.

--------------------------------

Propose a system and fill in how it works for each user story.

**Comment by eidolonFIRE** [comment]
# Proposal:

## Register / Login
`Register` is called by client only once the first time a device hits the server.
1. User inputs basic information and client generates a `pub_id`
2. Client sends `pub_id` and info to server.
2. Server generates a `secret_id` and sends it back to client. (this is like a password)

Once the client remembers it's own `secret_id`, it instead calls `Login` each time a connection to the server is made. Providing the correct `pub_id` and matching `secret_id` grants access to the API via that socket.

## Transfer login / desktop access
Couple layers down in menu is option to view your `pub_id` and `secret_id`.
Advanced option to override your generated ID with another. This way you can manually copy your profile from one devices to another by using your `secret_id`
This could later be upgrade with a QR code but it's a very rare case. Also, if it's possible to clone a profile, we'll have to resolve problem when you login multiple times with the same profile. `(multiple devices with same profile can't operate at the same time)`.

Desktop access would simply prompt visitor for `pub_id` of the pilot to be viewed. This could be saved as a cookie for quick access in the future.

## Meetups (aka Groups)
A meetup is a glob of pilots that are presumably flying together.
The meetup services (messaging, telemetry, map layers, etc) last as long as more than two pilots remain in the meetup.

Joining on any pilot in the meetup will add you to the glob. You can opt to leave the meetup. Or, if the app is closed for `4 hours` without interacting with the meetup, it will not attempt to rejoin it next time it connects to the server. Meetups will fizzle out by natural attrition after a flight. Or, if flying multiple hops, the user activity will act as a heart-beat keeping the group together. If a user gets bumped out prematurely, just join back on another pilot who is still in the group.

## Contacts / Discoverability
App local storage will accumulate list of pilots you have flown with before.
To make a first contact when you don't get have contacts for anyone in a group you are joining, have one of them hold up QR code for their ID. Use your phone camera to view the QR and boom, you're in the group. (this uses `pub_id`)

(It would be interesting to use a proximity based auto group finding (server side) but I can't think of how to secure it so it couldn't be abused. It would sacrifice pilot privacy. )

## XC flight leader
There is no explicit flight leader. Any user can add/remove their own map layers which are displayed to the whole meetup. When a user leaves the meetup, their content stays, and anyone can remove it.
Each pilot selects their own waypoints from the shared list of waypoints.
In chat, any user can propose a next waypoint. Another pilot viewing the message can click it and automatically set their next waypoint as well.

## Spectating
Requires the same "login" procedure for a standard user.
Menu option to enter "spectating mode". No telemetry sent to server, but still have full access to view others and enter chat.

> `Caleb` is going to `Backup LZ`. Click to follow.

## Splitting off / Buddy system
There's no currently no "buddy system". Although, you could maybe mark other pilots as "VIP" on your own overlay to keep tabs on them?

When leaving a group manually, you can choose to "Split Group". Every user in the meetup is offered choice to follow the splitting off user.

-------------------------------------------
# User Stories

### A) Impromptu joining another pilot or group at the field.
If I know the pilot, just select them from my address book. If I don't, share my QR code or they share theirs. We form a meetup.
My phone will automatically leave our little meetup after `4 hours` of inactivity (or until the other user(s) have left).

### B) Planning XC flight days in advance.
I prepare my waypoints in google docs map. After I join the meetup, I click to import the layers / waypoints  from my document.

### C) Multi-hop, all-day XC.
After each hop, I remove the map layers that I added, and add map layer for the next segment. Our constant activity  keeps the meetup active. If someone drops, they just join on one of us and are back in.

### D) Getting new device / Accessing profile from desktop.
( covered above )

### E) Spectating a flight.
( covered above )

**Reply by eidolonFIRE** [reply]
^ PR for this proposal https://github.com/eidolonFIRE/xcNav_protocol/pull/1

**Reply by gitToad** [reply]
I want to understand your concern for **user privacy** (I am fully onboard with privacy, but for purposes of this discussion, we likely want a user name at least first name and maybe an icon picture, and maybe in future as a pilot choice, their wing so you can keep track of what you fly with/different fuel consumptions) We are not collecting other PID - I don't think we need email, we might want phone nr so we can allow "click to call" as a simple operation in the UI, but that providing that would be optional. Is there a privacy aspect I am overlooking ? 
Perhaps security, ie someone hacking into our server and downloading our phone numbers ?

**Reply by gitToad** [reply]
**_Starting_ a flight without having internet access**

It sometimes happens (like on the first flight on the SF to Salton) that the LZ is in an internet dead area. Now maybe some pilots already joined the meetup via a URL on slack.whatsapp.email etc. but if not, we should still allow a meetup to form ad hoc ? If we use a meetup GUID that can be generated without server interaction and shared locally at the LZ, then assuming you have coverage once you take off you can still get full group functionality.

**Flights without any internet**

Solo or as a group. You can still record telemetry locally on the phone, and assuming you have QR-code shared the meetup GUID prior to take off we should be able to record all these flights to the server once you have connectivity again (even post flight, a day later etc.) and because all the flights share the same meetup, record as  a single meetup. So telemetry and flights and meetup information are not lost even when there is no connection while flying.

**Reply by eidolonFIRE** [reply]
**Privacy**
I'm less concerned about user privacy to us, more concerned about security and data leaking. If the API is poorly setup, a malicious user could scrape all the data of all users or grief random meetups. For instance, if we offered gps based proximity auto joining, you could fake your location in various locations to join up with groups you aren't a part of.

By all new users having a sponsor and everything working as "friend of a friend" type invitations, the social networks of pilots are naturally cordoned off to trolls. You initially need a physical QR code invitation to get in and make your first contacts. I dunno if it'd always be this way, but for starters, it'd be a controlled way to grow the user base. If an invitation was abused and used to let in 100's of trolls, we'd be able to track that.

As for collecting user data such as phone number and email... if we do that, we need to really know what we're doing with data retention and security. That can get very serious if we get hacked and all the data is scraped. It's my preference to set this up such that our server never directly holds raw user data, so we never need to worry.

**Flights without internet**
I'll be thinking about this one... Our principles don't explicitly address these use cases you proposed. https://github.com/eidolonFIRE/xcNav/wiki/App-Principles They don't really support them either tho. Most of the emphasis is on in-flight and at-field priority. Recording your own telemetry is a given, but the bit about assembling the group after the flight is a tricky one. I definitely don't want to require every meeting to have an explicit code, but there's also no reason you couldn't opt to share one. This may just be a matter of adding API to upload old telemetry because atm telemetry is only sent in real-time.
